### DISCUSSION TITLE
Pre-allocate task IDs when creating chain

### DISCUSSION THREAD

**Discussion by WilliamDEdwards** [celery/celery#7299] [Q&A]
Hi,

I'd like to run some tasks in asynchronously (as in: one after the other), and immediately return the task IDs. The use case is to be able to do some stuff outside of the HTTP request-response cycle, and allow users to poll for the task status with task IDs.

While playing with `chain`, I noticed that the tasks aren't actually created until they are called (i.e. the previous task in the chain succeeded). This makes sense, but means I have to wait until the chain is finished to get all task IDs. 

Is there a (generally accepted) way to create the tasks in advance, keep them in a pending state, and call them or revoke them if a task in the chain doesn't succeed?

**Comment by thedrow** [comment]
Do `my_chain.freeze()` to freeze all the task ids.

**Reply by WilliamDEdwards** [reply]
Thanks for your reply.

I looked at the Canvas code. `freeze` seems to set a UUID for a signature before running it. (The name isn't very intuitive; it seems to be Celery internal lingo :) )

Calling `freeze` on each task in the chain before running it, does seem to be what I'm looking for:

```
for i in tasks:
    print(i.freeze().id)
```

... where `tasks` is a list with `celery.canvas.Signature`s, yields:

```
1e475cf3-14a3-408b-92d3-e3804a1615c1
0bc648b8-8066-42d9-8b55-64de91dce4ed
4602c022-25f7-48dd-b486-7c196779bfac
```

Maybe calling `freeze` on the chain would work as well, haven't tested that yet. I don't see any looping logic in the code, though.

**Reply by thedrow** [reply]
Calling freeze on the chain should freeze all tasks.

**Reply by WilliamDEdwards** [reply]
Clear. The documentation isn't very clear on this subject, so I'm not 100% sure how I should get the task IDs after freezing.

This seems to work:

```
chain_.freeze()

for task in chain_.tasks:
  print(task.id)
```

Is this the generally accepted way to do it?

**Comment by WilliamDEdwards** [comment]
Turns out this is somewhat a duplicate of https://github.com/celery/celery/issues/1813.
